sched/semaphore/sem_holder.c

```diff
@@ -1035,18 +944,57 @@ void nxsem_release_holder(FAR sem_t *sem)
 {
   FAR struct tcb_s *rtcb = this_task();
   FAR struct semholder_s *pholder;
+  FAR struct semholder_s *candidate = NULL;
+  unsigned int total = 0;
```

Comment:
In general I would suggest to rewrite logic in the next way:
```
void nxsem_release_holder(FAR sem_t *sem)
{
  FAR struct tcb_s *rtcb = this_task();
  FAR struct semholder_s *pholder;
  FAR struct semholder_s *candidate = NULL;

  /* Find the container for this holder */

#if CONFIG_SEM_PREALLOCHOLDERS > 0
  for (pholder = sem->hhead; pholder; pholder = pholder->flink)
#else
  int i;

  /* We have two hard-allocated holder structures in sem_t */

  for (i = 0, pholder = &sem->holder[0]; i < 2; pholder = &sem->holder[i++])
#endif
  {
      if (pholder->htcb == rtcb)
        {
          candidate = pholder;
          break;
        }
      else if (pholder->htcb != NULL)
        {
          if ((candidate == NULL) || (candidate->counts < pholder->counts))
          {
            candidate = pholder;
          }
        }
  }

  /* We either found the task that is a holder or the best candidate */

  if (candidate != NULL)
  {
    /* Decrement the counts on this holder -- the holder will be freed
     * later in nxsem_restore_baseprio.
     */

    candidate->counts--;
  }  
}
```

Reply:
If there are 2 holders (not the rtcb) with the same counts which is 1, whoever is chosen will break the priority inheritance.
If neither pholder->htcb == rtcb nor total == 1 is met, it should be the wrong usage of the semaphore. Currently do nothing, as the original code. Leave it for later for us to find a smarter way, or change the design.

Reply:
Ok.